tests/test_api/test_validators/test_generate_teams_validator/test_generate_teams_validator.py

```diff
@@ -0,0 +1,219 @@
+                                "attribute": 4,
+                                "operator": "more than",
+                                "value": 0,
+                            },
+                        ],
+                    },
+                    {
+                        "id": 2,
+                        "project_id": 2,
+                    },
+                ],
+                "total_teams": 2,
+                "max_team_size": 2,
+                "min_team_size": 1,
+            },
+        }
+
+        cls.valid_data = valid_data.copy()
+
+        missing_students_data = valid_data.copy()
+        del missing_students_data["students"]
+        cls.missing_students_data = missing_students_data
+
+        missing_algorithm_options_data = valid_data.copy()
+        del missing_algorithm_options_data["algorithm_options"]
+        cls.missing_algorithm_options_data = missing_algorithm_options_data
+
+        missing_team_generation_options_data = valid_data.copy()
+        del missing_team_generation_options_data["team_generation_options"]
+        cls.missing_team_generation_options_data = missing_team_generation_options_data
+
+        student_with_invalid_relationship_data = copy.deepcopy(valid_data)
+        student_with_invalid_relationship_data["students"][0]["relationships"] = {
+            "-1": "enemy"
+        }
+        cls.student_with_invalid_relationship_data = (
+            student_with_invalid_relationship_data
+        )
+
+        student_with_extra_attribute_data = copy.deepcopy(valid_data)
+        student_with_extra_attribute_data["students"][0]["attributes"]["5"] = [1, 2, 3]
+        cls.student_with_extra_attribute_data = student_with_extra_attribute_data
+
+        student_with_invalid_project_id_data = copy.deepcopy(valid_data)
+        student_with_invalid_project_id_data["students"][0][
+            "project_preferences"
+        ].append(-1)
+        cls.student_with_invalid_project_id_data = student_with_invalid_project_id_data
+
+        team_generation_options_with_duplicate_ids_data = copy.deepcopy(valid_data)
+        team_generation_options_with_duplicate_ids_data["team_generation_options"][
+            "initial_teams"
+        ] += [
+            {"id": 3, "project_id": 3},
+            {"id": 3, "project_id": 4},
+        ]
+        cls.team_generation_options_with_duplicate_ids_data = (
+            team_generation_options_with_duplicate_ids_data
+        )
+
+    def test_valid__happy_json(self):
+        try:
+            validator = GenerateTeamsValidator(self.valid_data)
+            validator.validate()
+        except SchemaError as e:
+            self.fail("Valid data raised SchemaError unexpectedly")
+
+    def test_errors__top_levels(self):
+        with self.assertRaises(SchemaError) as e:
+            validator = GenerateTeamsValidator(self.missing_students_data)
+            validator.validate()
+        self.assertEqual(e.exception.code, "Missing key: 'students'")
```

Comment:
don't remove now, but for the future idk if we should write tests that are this specific, but I know why you did it.

I feel like if you broke the validator class into smaller methods, the tests could also be smaller, and then you wouldnt need to test the exact match of the exception code, you could just test if individual validation methods raise errors